### ISSUE TITLE
NEW API Planning & Discussion thread

### ISSUE THREAD

**Issue by Kukks** [btcpayserver/btcpayserver#1320]
This year we'll be focusing on making that new api come to life. The new api should not just be about replacing the compatible Bitpay API, but to also allow full usage of all of BTCPay's features, without its UI.

Before we can actually start opening PRs to add these APIs, we need to merge an authentication/authorization method, to which I currently propose #1262 ( there is also OpenId support partially there but I consider it deprecated) 

In the meantime, I have a branch with various prototyping for new API endpoints that I've worked on for a while. https://github.com/Kukks/btcpayserver/tree/api-preview This branch contains endpoints to:
* CRUD stores
* CRUD rate rules for stores
* ~~CRUD bitpay api token for stores~~
* CRUD payment methods for stores
* CRUD Payment Requests

api dev doc preview: https://redocly.github.io/redoc/?url=https://gist.githubusercontent.com/Kukks/106d4c92058813204cad6dbefbe94058/raw/06a78f94bfc3922f3dd251c5a8df210df3de5c05/gistfile1.json

**Comment by dennisreimann** [comment]
I'd like to add a general health endpoint, which could be used to get the general status and high-level info like e.g. sync status. I'd be willing to work on this and help with the new API in general. :)

**Comment by Kixunil** [comment]
Why is OpenId deprecated? It seems extremely useful to me. It's an open standard which we could use to connect with other applications. I actually started researching this topic yesterday and plan to make great use of it in my project.

**Comment by Kukks** [comment]
> 
> 
> Why is OpenId deprecated? It seems extremely useful to me. It's an open standard which we could use to connect with other applications. I actually started researching this topic yesterday and plan to make great use of it in my project.

I stopped pushing for it after an extremely long and time consuming review process( which has not yet ended and finished being implemented after 1.5years). I created #1262  that covers most of the feature set we wanted with much less code and no external library dependencies.

**Comment by Kixunil** [comment]
Hmm, I'm not sure if that can support my use case. What I want:

* Just client - no OpenID/whatever provider needed
* Allow external application to manage users
* Allow external application to manage scopes
* Do not bother the user with btcpayserver login page

Would it be possible to implement it? Maybe using nginx auth_request?

**Comment by Kukks** [comment]
This should be fine, you just generate an api key and then use it with
every http request in the header.

**Comment by Kixunil** [comment]
It looks like the authorization is interactive, I'd like it to be non-interactive. Sign in at external SSO provider and go directly to btcpayserver, without other confirmations or having to have the account there in the first place (it should be created during first access).

**Comment by eltNEG** [comment]
I just got started with BTCPay and haven't looked at the code yet. But I would like to know if you would advise against building the API service as a separate docker service that can be attached to the current set of services. I would like to build the API service with golang as I am not familiar with c#

**Comment by Kukks** [comment]
> 
> 
> I just got started with BTCPay and haven't looked at the code yet. But I would like to know if you would advise against building the API service as a separate docker service that can be attached to the current set of services. I would like to build the API service with golang as I am not familiar with c#

That wouldn't work since you need the internal C# services to expose the logic, data and events correctly.

**Comment by eltNEG** [comment]
> > I just got started with BTCPay and haven't looked at the code yet. But I would like to know if you would advise against building the API service as a separate docker service that can be attached to the current set of services. I would like to build the API service with golang as I am not familiar with c#
> 
> That wouldn't work since you need the internal C# services to expose the logic, data and events correctly.

Ohk. Thanks

**Comment by Kixunil** [comment]
I just remembered that there are ways to specify the API declaratively and then auto-generate the code from that in any language. Could this approach be used here? Would help a lot other projects, which might want to use different langs.

**Comment by Kukks** [comment]
> 
> 
> I just remembered that there are ways to specify the API declaratively and then auto-generate the code from that in any language. Could this approach be used here? Would help a lot other projects, which might want to use different langs.

Yes, this is through Swagger/OpenAPI which I intend to use. :)

**Comment by Kixunil** [comment]
I'm very happy to hear that! Also @pavlenex you just made my day putting it to high priority, as I have some wild ideas on how I could use this and want to implement soonish. :)

**Comment by Kukks** [comment]
soonish is optimistic :)

**Comment by eltNEG** [comment]
I am also looking forward to it.

**Comment by Kixunil** [comment]
For me it means month or two. :) But even if not, not a huge deal.

**Comment by Kukks** [comment]
I've published a draft of some of the APIs I'm playing around with here
https://app.swaggerhub.com/apis/btcpayserver/BTCPayServer/1.0.0

I'm still experimenting with the documentation generator and nothing is final

**Comment by Kixunil** [comment]
I noticed amount in payment request is defined as decimal - is that float? Currency amounts must not be stored in floats because of accounting issues.

**Comment by Kukks** [comment]
> 
> 
> I noticed amount in payment request is defined as decimal - is that float? Currency amounts must not be stored in floats because of accounting issues.

No, https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types

**Comment by Kixunil** [comment]
Interesting, I wonder if there's native implementation of that in all langs via swagger.

**Comment by autotelik** [comment]
swagger looks good so far.  I'd like to be able to use BtcPay with multiple users, each with a store and wallet than can receive lightning payments. Basically an API that will enable micro payment use cases, supporting multiple providers of paid-for-content.  Pretty much as someone else described here, 

https://github.com/btcpayserver/btcpayserver/issues/1022

Thanks for all your amazing work.

**Comment by radWorx** [comment]
Is this meant to be a "Merchant" API dealing with Store/App specific calls? Or will it include a "Developer" API to interact with layers 1 & 2?

**Comment by Kukks** [comment]
> Is this meant to be a "Merchant" API dealing with Store/App specific calls? Or will it include a "Developer" API to interact with layers 1 & 2?

This is mostly meant for BTCPay functionality.

**Comment by NicolasDorier** [comment]
posting that here, I was thinking about how a cli would looks like:

![image](https://user-images.githubusercontent.com/3020646/74407289-062d6580-4e75-11ea-818c-89cf47daa472.png)

It may help to the design of the API.
I think we should start small.

**Comment by Kixunil** [comment]
@NicolasDorier please don't pass token via command line, that is insecure.
I'd suggest looking into a file (default: `~/.btcpayserver/Main/.cookie`), which can be overriden by `--cookie-file`
So the user can write `AUTH="--cookie-file /some/file"` making your example work securely.

**Comment by Kukks** [comment]
> 
> 
> @NicolasDorier please don't pass token via command line, that is insecure.
> I'd suggest looking into a file (default: `~/.btcpayserver/Main/.cookie`), which can be overriden by `--cookie-file`
> So the user can write `AUTH="--cookie-file /some/file"` making your example work securely.
> 
> Similarly, instead of `--password`, I'd suggest `--password-file` and maybe `--generate-passwoord` outputting a random password (trivial generating script: `head -c 18 /dev/urandom | base64`)

This is not how it works and cannot work that way. The key needs to be provided as it depends on the user( a user can generate multiple keys with different permissions).

**Comment by Kixunil** [comment]
OK, then don't use default, make `--token-file` a mandatory argument.

**Comment by Kukks** [comment]
It 

> 
> 
> OK, then don't use default, make `--token-file` a mandatory argument.

It is like that, nicolas just made a local var and added it in each call

**Comment by Kixunil** [comment]
I understand. What I'm saying is store the secret into a file, not directly at command line. Having a default file was just an idea to make it more convenient for common cases, but not related to security.

**Comment by haarts** [comment]
> > I noticed amount in payment request is defined as decimal - is that float? Currency amounts must not be stored in floats because of accounting issues.
> 
> No, https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types

While I cannot argue with the official docs I do believe the narrative changes when serializing this type to JSON. Different languages parse floats different because the JSON spec doesn't specific how this should be done. Sometimes you see floats in JSON documents as Strings because of this. I would argue that having only ints in any payload is preferred. 

(for example this morning I had to write `(double.parse(invoice['btcPaid']) * Coin.bitcoin().multiplier).round()`, this is in Dart which doesn't have a `decimal` type, many languages don't.)

**Comment by haarts** [comment]
I'm interested in this subject and if you don't mind I'll give my 2 cents here. Feel free to do with it whatever you want. I spent some time reading but certainly didn't touch upon every single endpoint. For example I don't know if it makes sense to update a payment request. Why not just create a new one and forgo all the complexities which come with this extra code path. Less is more.

I like what I'm seeing so far on https://app.swaggerhub.com/apis/btcpayserver/BTCPayServer/1.0.0. 
I like the `/api/v1` prefix. 
I like plural resources (like `stores`) in the paths. 
I like the sane nesting of resources (a store has a payment method).
I like the use of HTTP verbs. At some places a PUT could be replaced with PATCH. Common misconception.
I like the `preview` endpoints.

### Authentication

You're proposing using API keys for authentication. Dropping the request signing as it is currently done. I do think it is important to understand what request signing brings to the table and what we will consequently lose if we don't do it. 
IMO request signing prevents accidental leaks of secrets. With signing we don't send a secret so there's nothing to leak. With an API key we send the secret with each request. This is not a great problem when BTCPay Server is always serving ONLY TLS requests. But we can't guarantee that. People are free to install BTCPay Server anywhere. Also in a plain HTTP environment. Or perhaps they terminate the TLS connection at some load balancer and mess up their routing through a potentially insecure environment. We don't know. With request signing these problems do not exist. I would keep request signing. We're taking about a piece of software which is dealing with real money.

Unrelated; why not also have the GET request `/api-keys/authorize` under the `/api/v1` path?

### Misc

- Let each field be of a single type (I'm looking at you `exceptionStatus`! This field can hold a bool or a string).
- Each response should include all fields always. Don't omit/add fields depending on the situation.
- Send a JSON error, not a HTML page whenever an exception occurs. Always. Also on 500s.
- Use HTTP status codes. 201 is great. 302 is deprecated.
- Tell the API consumer what went wrong and why. Error messages are really important.

This already is quite the wall of text. :)

**Comment by gerbz** [comment]
Awesome - we've been wanting to allow BTCPay integration with BitLift!  Seems like with the API merchants could add an API key and Store ID to their BitLift profile so we can generate invoices from their server when a customer buys their products?

I see in the docs we can POST to _/api/v1/stores/{storeId}/payment-requests_, but I dont see payment address info in the response.  Am I misunderstanding how this would work?  We would need all the payment info to integrate into our UI as oppose to iFraming or redirecting to an invoice.

**Comment by kemoantemo** [comment]
Hi guys i think you remember that i was asking for a feature from the early days of our beloved project so i hope it will be added with the new api can we have a field that calculates all the confirmed balance that have been received? And if any new transaction became to the invoice it sends another api with the new confirmed balance so we don't need to handle the exception transactions manually and get the confirmed balance directly from api and use it in our programs it will be really great and sorry for my horrible English and i hope u can understand me ♥️💚

**Comment by Kukks** [comment]
> 
> 
> Hi guys i think you remember that i was asking for a feature from the early days of our beloved project so i hope it will be added with the new api can we have a field that calculates all the confirmed balance that have been received? And if any new transaction became to the invoice it sends another api with the new confirmed balance so we don't need to handle the exception transactions manually and get the confirmed balance directly from api and use it in our programs it will be really great and sorry for my horrible English and i hope u can understand me ♥️💚

This has been available in the current invoice api for a long time. When you call get invoice through the bitpay api, there is a property called `CryptoInfo` which is an array of payments by each payment method, and each of them has a `Payments` list which has a `Confirmed` boolean.

**Comment by Kukks** [comment]
> 
> 
> Awesome - we've been wanting to allow BTCPay integration with BitLift! Seems like with the API merchants could add an API key and Store ID to their BitLift profile so we can generate invoices from their server when a customer buys their products?
> 
> I see in the docs we can POST to _/api/v1/stores/{storeId}/payment-requests_, but I dont see payment address info in the response. Am I misunderstanding how this would work? We would need all the payment info to integrate into our UI as oppose to iFraming or redirecting to an invoice.

This was just a quick prototyping of how it would work, not final at all.

**Comment by Kukks** [comment]
>  I like the use of HTTP verbs. At some places a PUT could be replaced with PATCH. Common misconception.

Hey @haarts,
In regards to 
> I like the use of HTTP verbs. At some places a PUT could be replaced with PATCH. Common misconception.

From my understanding, using `PATCH` implies that you would apply changes to the model based on what is supplied while a `PUT` replaces the object. Would it then not be fair to use `PUT` to replace models that were returned through `GET` fully? 

> Unrelated; why not also have the GET request /api-keys/authorize under the /api/v1 path?

Because it's not really an API but simply loads a UI for an end-user to create an API key.

In terms of api keys over non-https, perhaps we'll add a mandatory check that the connection is secure (https/localhost/tor) in order to authorize api calls. What's your thought @NicolasDorier ?

**Comment by haarts** [comment]
Hi,

I had to look up the relevant [RFC](https://tools.ietf.org/html/rfc5789) again. I believe you're right and I can't reproduce what 'some places' are. So `PUT`: complete replacement (or creation) of the object, `PATCH`: partial update.

> Because it's not really an API but simply loads a UI for an end-user to create an API key.

Ah, I missed that part. That makes perfect sense.

Re API keys over non-https: there's nothing you can do. The mandatory check you are talking about happens at the server in which case the damage is done. The key is already transmitted in plain text. The tricky bit here is that you do not control the environment a BTCPayServer is deployed in. Just not accepting plain API keys as authentication gets round this problem. This is of course not entirely without trade offs. Making a client which signs requests is more tedious than just chucking an API key in a header/request.

**Comment by Kixunil** [comment]
@haarts you're correct that the damage is already done. It can still have value as it's expected that the user setting it up will test it right after setting up and the failing test will notify him of a problem. I'd also suggest complete invalidation of the API token (so he can't just add HTTPS thinking it's fine).

Request signing is obviously more robust. Tediousness can be solved by providing libraries.

**Comment by Kukks** [comment]
> @haarts you're correct that the damage is already done. It can still have value as it's expected that the user setting it up will test it right after setting up and the failing test will notify him of a problem. I'd also suggest complete invalidation of the API token (so he can't just add HTTPS thinking it's fine).
> 
> Request signing is obviously more robust. Tediousness can be solved by providing libraries.

We have libraries for the current bitpay api in a variety of languages. Users are just equally frustrated with complexity of signing.

**Comment by Kixunil** [comment]
I find it very surprising, do you happen to have any resources on that? I suspect it's BitPays' fault.

I don't see any good reason why a well-design API should be harder with signing. The only difference should be that the function that takes API key as a parameter uses it differently under the hood, but that's not a concern for the user.

**Comment by Kukks** [comment]
> I find it very surprising, do you happen to have any resources on that? I suspect it's BitPays' fault.
> 
> I don't see any good reason why a well-design API should be harder with signing. The only difference should be that the function that takes API key as a parameter uses it differently under the hood, but that's not a concern for the user.

I deal with people trying to use the api directly or through the clients and they often complain of the complex auth system. Most are used to a simple auth header and don't want to deal with additional hashing library dependencies or a dedicated library for just some http requests. This probably comes up a few times a week.

**Comment by Kixunil** [comment]
Oh, I see, it makes sense then. A different approach would be to provide both authentication options, signing on by default. BTW, I find pubkey-based signing unnecessarily complex. HMAC-SHA256 should be sufficient.

There's a possible danger of signing: it may provide false sense of security, encouraging people to ignore TLS - it's still not confidential and a naive implementation doesn't solve replay attacks.

**Comment by MrAfrazKhan** [comment]
Hi, all I have integrated BTCPay Server for my system. One issue i am constantly facing is that i want to send money to other wallets through my own user interface not though BTCPay UI. For that i need APIs , so, it would be great if you people can expose send bitcoin from wallet API in your greenfield space ? Thanks

**Comment by Kixunil** [comment]
@MrAfrazKhan I think the wallet is actually handled by NBXplorer, so maybe you can already use its API to do the job?

**Comment by MrAfrazKhan** [comment]
@Kixunil but wallets are managed by BTCPay, how can I do that independently without interacting with the BTCPay server. If it's possible, then please tell me, how to do that?

**Comment by dennisreimann** [comment]
@MrAfrazKhan are you talking about a hot wallet you generated with BTCPay Server or is it an external wallet for which you imported the xpub?

**Comment by Kukks** [comment]
Looking for feedback on a new Invoice API for GreenField. Here is the PR #1760

**Comment by danielalexiuc** [comment]
This part of the Greenfield API is brilliant. Can you do something similar but for the Internal Bitcoin Node? 

i.e. getaddress / send / gettransactions / 

![image](https://user-images.githubusercontent.com/139083/94101233-d581c580-fe72-11ea-9884-3320388afa4a.png)

**Comment by alexgleason** [comment]
Hi, I'm implementing Greenfield API. One feature that seems to be missing is the ability to set the redirect URL when creating an invoice. Without it, the modal is the only viable option.
